server/nri-api.go

```diff
@@ -0,0 +1,741 @@
+// CRI (container and pod discovery, container adjustment and updates).
+//
+
+const (
+	nriDomain = "k8s.io"
+)
+
+func (a *nriAPI) GetName() string {
+	return nriDomain
+}
+
+func (a *nriAPI) ListPodSandboxes() []nri.PodSandbox {
+	pods := []nri.PodSandbox{}
+	for _, pod := range a.cri.ContainerServer.ListSandboxes() {
+		if pod.Created() {
+			pods = append(pods, nriPodSandbox(pod))
+		}
+	}
+	return pods
+}
+
+func (a *nriAPI) ListContainers() []nri.Container {
+	containers := []nri.Container{}
+	ctrList, err := a.cri.ContainerServer.ListContainers()
+	if err != nil {
+		log.Warnf(context.TODO(), "Failed to list containers: %v", err)
+	}
+	for _, ctr := range ctrList {
+		switch ctr.State().Status {
+		case oci.ContainerStateCreated, oci.ContainerStateRunning, oci.ContainerStatePaused:
+			containers = append(containers, &criContainer{
+				api: a,
+				ctr: ctr,
+			})
+		}
+	}
+	return containers
+}
+
+func (a *nriAPI) GetPodSandbox(id string) (nri.PodSandbox, bool) {
+	sandboxID, err := a.cri.PodIDIndex().Get(id)
+	if err != nil {
+		return nil, false
+	}
+
+	pod := a.cri.ContainerServer.GetSandbox(sandboxID)
+	if pod == nil {
+		return nil, false
+	}
+
+	return nriPodSandbox(pod), true
+}
+
+func (a *nriAPI) GetContainer(id string) (nri.Container, bool) {
+	ctr, err := a.cri.GetContainerFromShortID(todoMissingContext(), id)
+	if err != nil {
+		return nil, false
+	}
+
+	return &criContainer{
+		api: a,
+		ctr: ctr,
+	}, true
+}
+
+func (a *nriAPI) UpdateContainer(ctx context.Context, u *api.ContainerUpdate) error {
+	ctr, err := a.cri.GetContainerFromShortID(todoMissingContext(), u.ContainerId)
+	if err != nil {
+		// We blindly assume container with given ID not found and ignore it.
+		log.Errorf(ctx, "Failed to update CRI container %q: %v", u.ContainerId, err)
+		return nil
+	}
+
+	state := ctr.State().Status
+	if state != oci.ContainerStateRunning && state != oci.ContainerStateCreated {
+		return nil
+	}
+
+	resources := u.Linux.Resources.ToOCI()
+	err = a.cri.Runtime().UpdateContainer(ctx, ctr, resources)
+	if err != nil {
+		log.Errorf(ctx, "Failed to update CRI container %q: %v", u.ContainerId, err)
+		if u.IgnoreFailure {
+			return nil
+		}
+		return fmt.Errorf("failed to update CRI container %q: %w", u.ContainerId, err)
+	}
+
+	a.cri.UpdateContainerLinuxResources(ctr, resources)
+
+	return nil
+}
+
+func (a *nriAPI) EvictContainer(ctx context.Context, e *api.ContainerEviction) error {
+	ctr, err := a.cri.GetContainerFromShortID(todoMissingContext(), e.ContainerId)
+	if err != nil {
+		// We blindly assume container with given ID not found and ignore it.
+		log.Errorf(ctx, "Failed to evict CRI container %q: %v", e.ContainerId, err)
+		return nil
+	}
+	err = a.cri.stopContainer(ctx, ctr, 0)
+	if err != nil {
+		log.Errorf(ctx, "Failed to evict CRI container %q: %v", e.ContainerId, err)
+		return err
+	}
+
+	return nil
+}
+
+func todoMissingContext() context.Context {
```

Comment:
why don't we use context.TODO() instead of this?

Reply:
Good point. Changed to that.